### PR TITLE
Fast ICM42xxx performance

### PR DESCRIPTION
**Pull request by andyp1per**
At faster gyro backend rates (e.g. 4khz) the per-tick CPU load becomes quite important. This change reduces the CPU load a couple of ways:

- Increases the SPI clock on the MatekH743 to 16Mhz for the ICM42xxx gyros which are rated at up to 24Mhz (cannot currently get higher than 12.5Mhz because of the clock divisor). This gives about a 30% reduction in per-tick time.
- Synchronizes the ticks with the data. Before there were a lot of redundant ticks because data wasn't quite ready. Now the next tick is calculated from the time at which data is received.

### CODE REVIEW

**Comment by timtuxworth** [comment]
Would this also apply for the Zealot H743 which has a ICM-42688-P?

**Comment by andyp1per** [comment]
> Would this also apply for the Zealot H743 which has a ICM-42688-P

Yes, but manufacturer would need to confirm it is safe

**Comment by timtuxworth** [comment]
> > Would this also apply for the Zealot H743 which has a ICM-42688-P
> 
> Yes, but manufacturer would need to confirm it is safe

I've checked with QIo-Tek, they say this fix isn't necessary because their design doesn't have the problem this is fixing.

**Comment by andyp1per** [comment]
> I've checked with QIo-Tek, they say this fix isn't necessary because their design doesn't have the problem this is fixing.

Which can only mean they are not using this chip

**Comment by QioTek** [comment]
> > I've checked with QIo-Tek, they say this fix isn't necessary because their design doesn't have the problem this is fixing.
> 
> Which can only mean they are not using this chip

It needs to be tested. Will higher sampling make the sampling accuracy of the sensor different? We haven't conducted actual hardware testing yet. We need to compare with the previous sampling speed.

**Comment by andyp1per** [comment]
> It needs to be tested. Will higher sampling make the sampling accuracy of the sensor different? We haven't conducted actual hardware testing yet. We need to compare with the previous sampling speed.

For the 42688 the sampling is all done at 32Khz regardless of ODR. Higher ODR allows higher AP filter settings. This issue is addressing some issues that might be seen with higher ODR - mostly because the tick latency becomes more important

**Comment by andyp1per** [comment]
Flew four batteries on this today - all worked great

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp [hunk 1]
```diff
@@ -119,7 +119,7 @@ struct PACKED FIFOData {
 };
 
 #define INV3_SAMPLE_SIZE sizeof(FIFOData)
-#define INV3_FIFO_BUFFER_LEN 8
+#define INV3_FIFO_BUFFER_LEN 2
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp:122]: i'd be worried about getting behind

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp]: Fixed

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
tested on Pixhawk6X, works well!

**Pull request merged**